### DISCUSSION TITLE
Enhancement: Allow ravedude to force the AVR to reset.

### DISCUSSION THREAD

**Discussion by ysterbal** [Rahix/avr-hal#485] [Ideas]
After falling over this a couple of times, I believe it might be a good idea to make things a little simpler for flashing to AVR when it requires a hard reset, but there is no explicit reset button, only a hard reboot. 

The Adafruit ItsyBitsy 32u4-5V 16MHz board has a reset, but it uses the double-tap interface to reset to programming mode. 

This meant I had to flash my binaries to the device manually jiggle the double press on the USB connection on both Linux and Windows and frankly, it's bloody annoying.

So I suggest we use the same approach the Arduino IDE uses:  Open the port with BAUD 1200, which forces the device into programming mode, and then wait for n seconds and run the avrdude command. 

In my testing, this works for most AVRs, but I would appreciate your thoughts before contributing my changes for this enhancement.

**Comment by stappersg** [comment]
On Sat, Dec 30, 2023 at 09:53:13AM -0800, Johan Van Zijl wrote:
> 
> So I suggest we use the same approach the Arduino IDE uses:  Open the
> port with BAUD 1200, which forces the device into programming mode,
> and then wait for n seconds and run the avrdude command.
 


> In my testing, this works for most AVRs, but I would appreciate your
> thoughts before contributing my changes for this enhancement.

Please describe the "trick" in more detail.

I'm asking because I don't understand that only waiting with
baudrate at 1200b/s on development host triggers programming mode.
(I mis something  like sending a character sequence (at 1200bd)).


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by ysterbal** [comment]
So basically if you look at the code used in the Arduino IDE for
programming the first thing it does is open the COM port with a 1200 baud
this seems to basically reset the the device but basically only a soft
reset, I am digging through the boot loader to understand it better right
now but it works.


Check out the code snippet here:
https://github.com/arduino/Arduino/blob/master/arduino-core/src/cc/arduino/packages/uploaders/SerialUploader.java

see line 130-136

Hope it helps.

**Reply by stappersg** [reply]
_"Oh chips!"_

**Reply by stappersg** [reply]
Below [`SerialUploader.java`, line 139](https://github.com/arduino/Arduino/blob/master/arduino-core/src/cc/arduino/packages/uploaders/SerialUploader.java#L139) are in _comments_ the ingredients of the reset trick:  The Watch Dog Timer, WDT and the [Data Terminal Ready, DTR, line](https://en.wikipedia.org/wiki/Data_Terminal_Ready).

Elaboration on the _"Oh chips!"_.  There is **hardware**, do **NOT** ignore it. Grow from

>  Open the port with BAUD 1200

to

> Toggle the DTR line, it is connected to the watch dog timer on the device


(to be continued)

**Reply by stappersg** [reply]
That _"DTR line is connected to the watch dog timer on the device"_ is only true for some boards. Most likely for "arduino" boards. But `avrdude` is **not** "arduino"!   And `ravedude` is a wrapper for `avrdude`, not for "arduino".

**Reply by Rahix** [reply]
> And ravedude is a wrapper for avrdude, not for "arduino".

Sorry, but this is wrong.  `ravedude` bridges the gap between `avrdude`, Rust, and specific AVR boards (_including_ Arduino Boards).  Otherwise, `ravedude` wouldn't take a board identifier like `uno` or `nano` to select the port and `avrdude` settings.

**Comment by ysterbal** [comment]
Understood for me, this was mainly around the Arduino, I have found the
others to work just fine.

That is why I wondered if we could support this as a feature.

I did figure out it was hardware in the end, but thnx for the clarification,
I might help others as well.

**Reply by ysterbal** [reply]
Just a note: this is actually dealt with in board.rs, and the only necessary thing is to add the ItsyBitsy 3v and 5v VID/PID 

### board.rs:
```  
fn guess_port(&self) -> Option<anyhow::Result<std::path::PathBuf>> {
        Some(find_port_from_vid_pid_list(&[
            (0x2341, 0x0036),
            (0x2341, 0x8036),
            (0x239A, 0x000E),  // Support for ItsyBitsy 5V
            (0x239A, 0x000D),  // Support for ItsyBitsy 3V
            (0x2A03, 0x0036),
            (0x2A03, 0x8036),
        ]))
    }
```

Unless others need this, I am just going to keep this in my fork, so please consider this dealt with.

**Reply by Rahix** [reply]
Ah right, the Arduino Leonardo board code already includes the relevant line.  @stappersg so it's crystal clear for you as well, we literally include the line

https://github.com/Rahix/avr-hal/blob/7fc791145f37df214086b769f1b76475d998f1f1/ravedude/src/board.rs#L174-L174

to do the "open port at 1200 baud".

---

> Unless others need this, I am just going to keep this in my fork, so please consider this dealt with.

No no, I do want this included in upstream ravedude please!  You certainly won't be the only one running into this.  For the time being, my suggestion would be to simply duplicate the `ArduinoLeonardo` board code for `ItsyBitsy` and make it match the two ItsyBitsy VID:PID pairs.  @ysterbal, would you be willing to send a pull-request for this?
